File: ss_listy.py

class Listy(object):
	def __init__(self, l):
		self.l = l

	def element_at(self, i):
		if i < len(self.l):
			return self.l[i]
		return -1

def bs_listy(listy, x):
	start = 0
	end = 0
	c = 0
	while listy.element_at(end) != -1 and listy.element_at(end) <= x:
		start = end
		end = pow(2, c) - 1
		c += 1
	print(start, end)
	while start <= end:
		middle = (start + end) / 2
		if listy.element_at(middle) == - 1 or listy.element_at(middle) > x:
			end = middle - 1
		elif listy.element_at(middle) < x:
			start = middle + 1
		else:
			return middle
	return -1

def main():
	listy = Listy([i for i in xrange(200)])
	print(bs_listy(listy, 199))



if __name__ == "__main__":
	main()

File: ss_listy.cpp
#include <iostream>
#include <math.h>

#define NULL_VALUE -1
#define SIZE 100

using namespace std;

class Listy{
public:
	Listy();
	~Listy();
	int elementAt(int);
	int l[SIZE];
};

Listy::Listy(){
	for(int i=0; i<SIZE; i++){
		l[i] = 2*i;
	}
}

int Listy::elementAt(int i){
	if(i < SIZE){
		return l[i];
	}
	return NULL_VALUE;
}

int bsListy(Listy* listy,int x){
	int start = 0;
	int end = 0;
	int c = 0;
	int middle = 0;
	while(listy->elementAt(end) != -1 && listy->elementAt(end) <= x){
		start = end;
		end = pow(2,c) - 1;
		c += 1;
	}
	while(start <= end){
		middle = (start + end)/2;
		if(listy->elementAt(middle) == -1 || listy->elementAt(middle) > x){
			end = middle - 1;
		}
		else if(listy->elementAt(middle) < x){
			start = middle + 1;
		}
		else{
			return middle;
		}
	}
	return NULL_VALUE;
}

int main(){

	Listy* listy = new Listy();
	cout << bsListy(listy, 20) << endl;

	return 0;
}
